src/library/scala/collection/immutable/NVector.scala

```diff
@@ -0,0 +1,2004 @@
+
+  protected[this] def slice0(lo: Int, hi: Int): NVector[A] = {
+    val b = new NVectorSliceBuilder(lo, hi)
+    b.consider(1, prefix1)
+    b.consider(2, prefix2)
+    b.consider(3, prefix3)
+    b.consider(4, data4)
+    b.consider(3, suffix3)
+    b.consider(2, suffix2)
+    b.consider(1, suffix1)
+    b.result()
+  }
+
+  override def tail: NVector[A] =
+    if(len1 > 1) copy(copyTail(prefix1), len1-1, len12 = len12-1, len123 = len123-1, length = length-1)
+    else slice0(1, length)
+
+  override def init: NVector[A] =
+    if(suffix1.length > 1) copy(suffix1 = copyInit(suffix1), length = length-1)
+    else slice0(0, length-1)
+
+  protected[immutable] def vectorSliceCount: Int = 7
+  protected[immutable] def vectorSlice(idx: Int): Array[_ <: AnyRef] = (idx: @switch) match {
+    case 0 => prefix1
+    case 1 => prefix2
+    case 2 => prefix3
+    case 3 => data4
+    case 4 => suffix3
+    case 5 => suffix2
+    case 6 => suffix1
+  }
+  protected[immutable] def vectorSlicePrefixLength(idx: Int): Int = (idx: @switch) match {
+    case 0 => len1
+    case 1 => len12
+    case 2 => len123
+    case 3 => len123 + data4.length*WIDTH3
+    case 4 => len123 + data4.length*WIDTH3 + suffix3.length*WIDTH2
+    case 5 => length - suffix1.length
+    case 6 => length
+  }
+
+  override protected[this] def appendedAll0[B >: A](suffix: collection.IterableOnce[B], k: Int): NVector[B] = {
+    val suffix1b = append1IfSpace(suffix1, suffix)
+    if(suffix1b ne null) copy(suffix1 = suffix1b, length = length-suffix1.length+suffix1b.length)
+    else super.appendedAll0(suffix, k)
+  }
+}
+
+
+/** 5-dimensional radix-balanced finger tree */
+private final class NVector5[+A](_prefix1: Arr1, len1: Int,
+                                 prefix2: Arr2, len12: Int,
+                                 prefix3: Arr3, len123: Int,
+                                 prefix4: Arr4, len1234: Int,
+                                 data5: Arr5,
+                                 suffix4: Arr4, suffix3: Arr3, suffix2: Arr2, suffix1: Arr1,
+                                 val length: Int) extends NVector[A](_prefix1) {
+  import NVectorStatics._
+
+  @inline private[this] def copy(prefix1: Arr1 = prefix1, len1: Int = len1,
+                                 prefix2: Arr2 = prefix2, len12: Int = len12,
+                                 prefix3: Arr3 = prefix3, len123: Int = len123,
+                                 prefix4: Arr4 = prefix4, len1234: Int = len1234,
+                                 data5: Arr5 = data5,
+                                 suffix4: Arr4 = suffix4, suffix3: Arr3 = suffix3, suffix2: Arr2 = suffix2, suffix1: Arr1 = suffix1,
+                                 length: Int = length) =
+    new NVector5(prefix1, len1, prefix2, len12, prefix3, len123, prefix4, len1234, data5, suffix4, suffix3, suffix2, suffix1, length)
+
+  @inline def apply(index: Int): A = {
+    if(index < 0 || index >= length) throw ioob(index)
+    if(index >= len1234) {
```

Comment:
Is it possible to get an encoding which does not have the potential of having to try N failed branches?

Reply:
No, all lengths are arbitrary. It's not a big issue though. Every additional branch is less likely by a factor of 16 (on average) and leads to a lookup with one indirection less (meaning the more branching you do first, the faster the final lookup will be).